File: apps/api/app/schemas/msme.py
from datetime import datetime

from pydantic import BaseModel, Field

from app.schemas.common import DocumentAvailability, Pagination, ScenarioLabel, Segment


class FinancialSnapshotSchema(BaseModel):
    snapshot_month: str
    monthly_revenue_avg: int
    monthly_expense_avg: int
    average_bank_balance: int
    cash_inflow_volatility: float
    revenue_growth_3m: float
    revenue_growth_6m: float
    emi_obligation: int
    existing_debt: int
    bounce_count_3m: int
    bounce_count_6m: int
    gst_filing_regularity: float
    buyer_concentration: float
    digital_payment_ratio: float
    gem_order_completion_rate: float | None = None
    invoice_delay_avg_days: int
    cash_deposit_ratio: float | None = None
    revenue_spike_ratio: float | None = None


class DocumentStatusSchema(BaseModel):
    bank_statement: DocumentAvailability
    gst_returns: DocumentAvailability
    udyam: DocumentAvailability
    bureau_report: DocumentAvailability
    itr: DocumentAvailability
    gem_profile: DocumentAvailability
    missing_documents: list[str]
    stale_documents: list[str]


class MSMEDetail(BaseModel):
    id: str
    business_name: str
    segment: Segment
    scenario_label: ScenarioLabel
    city: str
    state: str
    business_vintage_months: int
    employee_count: int
    requested_credit_amount: int
    financials: FinancialSnapshotSchema
    documents: DocumentStatusSchema
    latest_score_id: str | None = None
    latest_prospect_signal_id: str | None = None


class MSMEListItem(BaseModel):
    id: str
    business_name: str
    segment: Segment
    scenario_label: ScenarioLabel
    city: str
    state: str
    requested_credit_amount: int
    monthly_revenue_avg: int
    health_score: int | None = None
    risk_tier: str | None = None
    prospect_score: int | None = None
    prospect_priority: str | None = None
    data_confidence: int | None = None
    recommended_human_action: str | None = None


class MSMEListResponse(BaseModel):
    items: list[MSMEListItem]
    pagination: Pagination


class SeedRequest(BaseModel):
    reset: bool = True
    seed: int = 42
    profile_count: int = Field(default=9, ge=8, le=50)


class SeedResponse(BaseModel):
    seeded: bool
    profile_count: int
    scenario_counts: dict[str, int]
    audit_event_id: str
    generated_at: datetime

File: apps/web/src/lib/schemas/msme.ts
import { z } from "zod";
import { paginationSchema } from "./common";

export const financialSnapshotSchema = z.object({
  snapshot_month: z.string(),
  monthly_revenue_avg: z.number(),
  monthly_expense_avg: z.number(),
  average_bank_balance: z.number(),
  cash_inflow_volatility: z.number(),
  revenue_growth_3m: z.number(),
  revenue_growth_6m: z.number(),
  emi_obligation: z.number(),
  existing_debt: z.number(),
  bounce_count_3m: z.number(),
  bounce_count_6m: z.number(),
  gst_filing_regularity: z.number(),
  buyer_concentration: z.number(),
  digital_payment_ratio: z.number(),
  gem_order_completion_rate: z.number().nullable(),
  invoice_delay_avg_days: z.number(),
  cash_deposit_ratio: z.number().nullable(),
  revenue_spike_ratio: z.number().nullable()
});

export const documentStatusSchema = z.object({
  bank_statement: z.string(),
  gst_returns: z.string(),
  udyam: z.string(),
  bureau_report: z.string(),
  itr: z.string(),
  gem_profile: z.string(),
  missing_documents: z.array(z.string()),
  stale_documents: z.array(z.string())
});

export const msmeListItemSchema = z.object({
  id: z.string(),
  business_name: z.string(),
  segment: z.string(),
  scenario_label: z.string(),
  city: z.string(),
  state: z.string(),
  requested_credit_amount: z.number(),
  monthly_revenue_avg: z.number(),
  health_score: z.number().nullable(),
  risk_tier: z.string().nullable(),
  prospect_score: z.number().nullable(),
  prospect_priority: z.string().nullable(),
  data_confidence: z.number().nullable(),
  recommended_human_action: z.string().nullable()
});

export const msmeListResponseSchema = z.object({
  items: z.array(msmeListItemSchema),
  pagination: paginationSchema
});

export const msmeDetailSchema = z.object({
  id: z.string(),
  business_name: z.string(),
  segment: z.string(),
  scenario_label: z.string(),
  city: z.string(),
  state: z.string(),
  business_vintage_months: z.number(),
  employee_count: z.number(),
  requested_credit_amount: z.number(),
  financials: financialSnapshotSchema,
  documents: documentStatusSchema,
  latest_score_id: z.string().nullable(),
  latest_prospect_signal_id: z.string().nullable()
});

export type MSMEListItem = z.infer<typeof msmeListItemSchema>;
export type MSMEListResponse = z.infer<typeof msmeListResponseSchema>;
export type MSMEDetail = z.infer<typeof msmeDetailSchema>;
export type FinancialSnapshot = z.infer<typeof financialSnapshotSchema>;
export type DocumentStatus = z.infer<typeof documentStatusSchema>;
